test/docker/kafka/kafka_test.go

```diff
@@ -0,0 +1,217 @@
+package kafka
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/beatlabs/patron/component/async"
+
+	"github.com/Shopify/sarama"
+	"github.com/beatlabs/patron/log"
+	patrondocker "github.com/beatlabs/patron/test/docker"
+	"github.com/ory/dockertest"
+	"github.com/ory/dockertest/docker"
+)
+
+const (
+	kafkaHost     = "localhost"
+	kafkaPort     = "9092"
+	zookeeperPort = "2181"
+)
+
+func TestMain(m *testing.M) {
+	topics := []string{
+		getTopic(simpleTopic1),
+		getTopic(simpleTopic2),
+		getTopic(groupTopic1),
+		getTopic(groupTopic2),
+	}
+	k, err := create(60*time.Second, topics...)
+	if err != nil {
+		fmt.Printf("could not create kafka runtime: %v\n", err)
+		os.Exit(1)
+	}
+	err = k.setup()
+	if err != nil {
+		fmt.Printf("could not start containers: %v\n", err)
+		os.Exit(1)
+	}
+
+	exitCode := m.Run()
+
+	ee := k.Teardown()
+	if len(ee) > 0 {
+		for _, err := range ee {
+			fmt.Printf("could not tear down containers: %v\n", err)
+		}
+	}
+
+	os.Exit(exitCode)
+}
+
+type kafkaRuntime struct {
+	patrondocker.Runtime
+	topics []string
+}
+
+func create(expiration time.Duration, topics ...string) (*kafkaRuntime, error) {
+	br, err := patrondocker.NewRuntime(expiration)
+	if err != nil {
+		return nil, fmt.Errorf("could not create base runtime: %w", err)
+	}
+	return &kafkaRuntime{topics: topics, Runtime: *br}, nil
+}
+
+func (k *kafkaRuntime) setup() error {
+
+	var err error
+
+	runOptions := &dockertest.RunOptions{Repository: "wurstmeister/zookeeper",
+		PortBindings: map[docker.Port][]docker.PortBinding{
+			docker.Port(fmt.Sprintf("%s/tcp", zookeeperPort)): {{HostIP: "", HostPort: zookeeperPort}},
+			// port 22 is too generic to be used for the test
+			"29/tcp":   {{HostIP: "", HostPort: "22"}},
+			"2888/tcp": {{HostIP: "", HostPort: "2888"}},
+			"3888/tcp": {{HostIP: "", HostPort: "3888"}},
+		},
+	}
+	zookeeper, err := k.RunWithOptions(runOptions)
+	if err != nil {
+		return fmt.Errorf("could not start zookeeper: %w", err)
+	}
+
+	ip := zookeeper.Container.NetworkSettings.Networks["bridge"].IPAddress
+
+	kafkaTCPPort := fmt.Sprintf("%s/tcp", kafkaPort)
+
```

Comment:
Maybe lets add a comment to highlight the restrictions of this approach.
e.g. not able to run tests in parallel with multiple runtimes

Reply:
You can enable the tests to run in parallel since we open just one kafka instance. If you notice every test uses a different topic.
Locally it ran very nicely but in github it did not work probably because the instance there is not so powerful as my machine in order to run them fast enough...